components/ws-daemon/pkg/daemon/markunmount.go

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) 2021 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package daemon
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sys/unix"
+	"golang.org/x/xerrors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+
+	"github.com/gitpod-io/gitpod/common-go/log"
+	"github.com/gitpod-io/gitpod/ws-daemon/pkg/content"
+	"github.com/gitpod-io/gitpod/ws-daemon/pkg/dispatch"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	// propagationGracePeriod is the time we allow on top of a container's deletionGracePeriod
+	// to make sure the changes propagate on the data plane.
+	propagationGracePeriod = 10 * time.Second
+)
+
+// NewMarkUnmountFallback produces a new MarkUnmountFallback. reg can be nil
+func NewMarkUnmountFallback(reg prometheus.Registerer) (*MarkUnmountFallback, error) {
+	counter := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "markunmountfallback_active_total",
+		Help: "counts how often the mark unmount fallback was active",
+	}, []string{"successful"})
+	if reg != nil {
+		err := reg.Register(counter)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &MarkUnmountFallback{
+		activityCounter: counter,
+	}, nil
+}
+
+// MarkUnmountFallback works around the mount propagation of the ring1 FS mark mount.
+// When ws-daemon restarts runc propagates all rootfs mounts to ws-daemon's mount namespace.
+// This prevents proper unmounting of the mark mount, hence the rootfs of the workspace container.
+//
+// To work around this issue we wait pod.terminationGracePeriod + propagationGracePeriod and,
+// after which we attempt to unmount the mark mount.
+//
+// Some clusters might run an older version of containerd, for which we build this workaround.
+type MarkUnmountFallback struct {
+	mu      sync.Mutex
+	handled map[string]struct{}
+
+	activityCounter *prometheus.CounterVec
+}
+
+// WorkspaceAdded does nothing but implemented the dispatch.Listener interface
+func (c *MarkUnmountFallback) WorkspaceAdded(ctx context.Context, ws *dispatch.Workspace) error {
+	return nil
+}
+
+// WorkspaceUpdated gets called when a workspace pod is updated. For containers being deleted, we'll check
+// if they're still running after their terminationGracePeriod and if Kubernetes still knows about them.
+func (c *MarkUnmountFallback) WorkspaceUpdated(ctx context.Context, ws *dispatch.Workspace) error {
+	if ws.Pod.DeletionTimestamp == nil {
+		return nil
+	}
+
+	c.mu.Lock()
```

Comment:
What about
``
func(){
  c.mu.Lock()
  defer c.mu.Lock()
  ...
}()
```
as a pattern to indicate we really wan tto releace `c.mu`?

Reply:
Indeed we could write it that way. In this instance I found the `func` wrap too much considering it's just one line inbetween. A matter of taste I reckon :)

Reply:
I should have paid more attention as to where that comment is located 🤦 
Will update accordingly